kernel/per_cpu/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+///
+/// ## Required traits
+/// Each field in this struct must implement [`cpu_local::PerCpuField`],
+/// which in turn mandates that each field have a unique type distinct from the type
+/// of every other field.
+/// Currently we achieve this with newtype wrappers
+#[allow(dead_code)] // These fields are accessed via `cpu_local` functions.
+#[repr(C)]
+//
+// IMPORTANT NOTE:
+// * These fields must be kept in sync with `cpu_local::PerCpuField`.
+// * The same applies for the `const_assertions` module at the end of this file.
+//
+pub struct PerCpuData {
+    /// A pointer to the start of this struct in memory, similar to a TLS self pointer.
+    /// This has a different initial value for each CPU's data image, of course.
+    ///
+    /// We use this to allow writes to this entire structure (for initialization),
+    /// and also to allow faster access to large fields herein, as they don't need to be
+    /// loaded in full before accessing a single sub-field. See this for more:
+    /// <https://github.com/rust-osdev/x86_64/pull/257#issuecomment-849514649>.
+    self_ptr: usize,
+    /// The unique ID of this CPU.
+    cpu_id: CpuId,
+    /// The current preemption count of this CPU, which is used to determine
+    /// whether task switching can occur or not.
+    preemption_count: PreemptionCount,
+    /// A preemption guard used during task switching to ensure that one task switch
+    /// cannot interrupt (preempt) another task switch already in progress.
+    // task_switch_preemption_guard: Option<TestU32>, // TODO temp remove this
+    task_switch_preemption_guard: TaskSwitchPreemptionGuard,
+    /// Data that should be dropped after switching away from a task that has exited.
+    /// Currently, this contains the previous task's `TaskRef` that was removed
+    /// from its TLS area during the last task switch away from it.
+    drop_after_task_switch: DropAfterTaskSwitch,
+    test_value: TestValue,
+    test_string: TestString,
+}
+
+impl PerCpuData {
+    /// Defines the initial values of each per-CPU state.
+    fn new(self_ptr: usize, cpu_id: cpu::CpuId) -> Self {
+        Self {
+            self_ptr,
+            cpu_id: CpuId(cpu_id),
+            preemption_count: PreemptionCount(preemption::PreemptionCount::new()),
+            task_switch_preemption_guard: TaskSwitchPreemptionGuard(None),
+            drop_after_task_switch: DropAfterTaskSwitch(None),
+            test_value: TestValue(0xDEADBEEF),
+            test_string: TestString(alloc::string::String::from("test_string hello")),
+
+        }
+    }
+}
+
+define_per_cpu_field!(CpuId, cpu::CpuId, PerCpuField::CpuId);
+define_per_cpu_field!(PreemptionCount, preemption::PreemptionCount, PerCpuField::PreemptionCount);
+define_per_cpu_field!(TaskSwitchPreemptionGuard, Option<preemption::PreemptionGuard>, PerCpuField::TaskSwitchPreemptionGuard);
+define_per_cpu_field!(DropAfterTaskSwitch, Option<TaskRef>, PerCpuField::DropAfterTaskSwitch);
```

Comment:
AFAICT, this creates a circular dependency that `per_cpu` was trying to avoid in the first place.

`preemption` will not be able to access the CPU local `PreemptionCount` as that would lead to a circular dependency (`per_cpu` depends on `preemption` depends on `per_cpu`). The trait has to be implemented in the crate where the type is defined (e.g. in `cpu` or `preemption`).

Reply:
Yeah, that's true.

`preemption` is a special case, given that `cpu_local` itself must depend on `preemption` because its `CpuLocal::with()` accessors use `PreemptionGuard` and `hold_preemption()`. So in the next PR when we start using per-CPU preemption counts, I will have to merge `cpu_local` and `preemption`. That's a known issue, so no worries there.

I'll work on this.
```rust

```